### ISSUE TITLE
Window movement while starting neovim

### ISSUE THREAD

**Issue by lf-** [mintty/mintty#629]
When I ssh into my linux box, then start `nvim` when mintty is maximised, the mintty window moves right and the title bar gets slightly taller. I can reproduce this with 100% reliability.

Strangely, this isn't reproduced with the vim-minimal package (vim 8.0).

![bug1](https://cloud.githubusercontent.com/assets/6652840/22871809/2aefe3f8-f17a-11e6-92b4-b1b32ab828ca.PNG)

Look at the left part of the window. All that was done was that I started neovim.

![bug2](https://cloud.githubusercontent.com/assets/6652840/22871811/2da43360-f17a-11e6-9996-2edbf94500e5.PNG)

**Comment by mintty** [comment]
This must be caused by an escape sequence and we should find out which.
Please make a screen log (e.g. starting `mintty -l log629`) and stop it (e.g. copy the file) immediately after the effect. (With the current repository version you could make more specific screen logs, toggling from the context menu.)

**Comment by lf-** [comment]
[log629.txt](https://github.com/mintty/mintty/files/776014/log629.txt)

Here's a log.

**Comment by lf-** [comment]
[log629-2.txt](https://github.com/mintty/mintty/files/776018/log629-2.txt)

Here's another log, this time with vim (vim-minimal fedora package). This one *doesn't* cause the problem. Since neovim is a fork of vim, it would likely be largely the same except for the problem portion... we can hope.

**Comment by mintty** [comment]
nvim sends the escape sequence [8;85;363t which sets the terminal size to 85 rows and 363 columns,
implicitly exiting full-screen or maximised mode. On that occasion, the window size is recalculated and passed to Windows. The repositioning of the left border is probably caused by Windows; there has been a similar effect (#599) but only in a Windows developers preview release – do you run any of them? The resizing of the window title area seems to be an erratic side-effect of Windows resizing, it does not happen every time.

**Comment by lf-** [comment]
Yes, I am running Win10 1607, build 14986.1001 (Insiders).

**Comment by lf-** [comment]
"It does not happen every time" This bug does happen 100% of the time.

**Comment by mintty** [comment]
Occasionally, the window stay in fullscreen mode here, rather than setting the window size, which is another issue (and perhaps a bug).
About the left border gap, however, I can't do anything, it's a Windows bug.

**Comment by xelra** [comment]
In my #599 issue, I wasn't using a Windows developer's preview release. For me this bug happens 100% of the time too.
It doesn't happen in PuTTY though. So I don't think it's a Windows bug. Well, maybe. But PuTTY seems to do something different.

**Comment by xelra** [comment]
My personal guess is this: There's a fixed point which actually stays valid. Now if I drag the window to one of the display borders to be resized, it will resize the window to a size that is not an integer of rows and columns. This imperfection is inside of the window though. so it goes pretty much unnoticed.
When the window resizes itself though, it will resize to exact integers of rows and columns and then the remainder is outside of the window.

**Comment by mintty** [comment]
In #599, we had traced that mintty requests position 0 from Windows. Also that code has not changed recently. It may be that putty does something different, but I don't think I'll be analysing that. If someone finds out, it might give a clue for a workaround.
Not sure I understand your second comment properly. If mintty is resized and rounds to a multiple of column width (plus padding), it changes its width, not its left position.
Also, as noted above, the effect is caused by the sequence ESC[8;85;363t ; please try to send that separately (`echo -ne '\e[8;85;363t'`) from both mintty and putty.

**Comment by xelra** [comment]
Well, I thought that the fixed point might be in the top right corner, instead of the top left.

**Comment by xelra** [comment]
I've tested the sequence in Mintty and PuTTY. In Mintty it changes both position and size of the window and the position has a gap.
In PuTTY, the top left position isn't changed at all. It just resizes the window towards the bottom right corner.

**Comment by mintty** [comment]
Here putty does not move towards the upper left corner at all. It just stays where it is and expands to the right and bottom (beyond screen borders). Not really better. (This is cygwin putty.)

**Comment by mintty** [comment]
Intending to close as there is no evidence of a mintty bug.

**Comment by xelra** [comment]
Well, it's hard to tell whether it's a Windows or a mintty bug.

But since other terminals like PuTTY or ConEmu/Cmder are not showing this behavior, I'd lean towards it being a mintty bug. Or they've incorporated a workaround and it's a Windows bug.

The strangest thing though is that when the window is full screen and moves, it still thinks it's full screen. Because grabbing it from the top, will make it resume it's small window mode again.

**Comment by mintty** [comment]
For the records, we are talking about the effect of `echo -e '\e[8;85;363t'`.
From a full-screen window, it stays full-screen (but may remove the scrollbar or add 1 bottom line).
From a maximized window, it has no effect (I'm not sure whether it used to have one, but not anymore).
From a normal, smaller window, it has the effect of approximately maximizing the window, leaving a gap as described. However, the resulting terminal size is much smaller than the requested which is too large for the screen, so exact behaviour cannot be expected. I'll check whether the way of adjusting screen size can be tweaked in a convincing way, but I'd say this is a very minor issue.

Not sure what you mean with the last remark. In full-screen mode (Alt+Enter), there is no window title to grab, so the window cannot be moved.

If you claim we still have a significant issue, please provide a clear and reproducible description of any observed effect, also with other terminals to compare.

**Comment by xelra** [comment]
I'm not sure about the escape sequence, but I was testing with Neovim. The behavior from Neovim and the escape sequence actually differs.

In my earlier comment I meant maximized, not fullscreen.

I'll put together a detailed description of the differences between the terminal emulators mentioned, when starting Neovim.

**Comment by mintty** [comment]
In order to factor out differing eI'd appreciate if you extract an escape sequence to produce the respective effects, rather than refer to a tool that we may not share, esp. as you use it remotely.

**Comment by mintty** [comment]
In order to factor out the impact of differing environments, I'd appreciate if you extract an escape sequence to produce the respective effects, rather than refer to a tool that we may not share, esp. as you use it remotely.

**Comment by xelra** [comment]
I've made a series of pictures to illustrate the issue better. For each terminal emulator I've created screenshots for maximized and half-screen-maximized windows. In each one I've ssh'ed into a remote machine and started Neovim.

This is PuTTY:
![putty_fs](https://cloud.githubusercontent.com/assets/908568/23827394/51ab72fc-06b3-11e7-8354-4495a4329f32.png)
![putty_hs](https://cloud.githubusercontent.com/assets/908568/23827395/5420096c-06b3-11e7-939d-233e5ac8279a.png)

This is Cmder (a ConEmu console):
![cmder_fs](https://cloud.githubusercontent.com/assets/908568/23827404/6d445146-06b3-11e7-9d79-375b3cfdea66.png)
![cmder_hs](https://cloud.githubusercontent.com/assets/908568/23827406/6ef687c0-06b3-11e7-8675-1fe2dbd015ef.png)

This is mintty:
![mintty_fs](https://cloud.githubusercontent.com/assets/908568/23827414/777dc692-06b3-11e7-8106-09b1a8704753.png)
![mintty_hs](https://cloud.githubusercontent.com/assets/908568/23827415/79924660-06b3-11e7-9fc8-8128ffaefeb2.png)

As one can see from the screenshots, the issue only appears in mintty. On the maximized view, the window grows and shifts to the right. On the half-screen-maximized view, the window shrinks and shifts to the right.

Unfortunately I don't know how to debug this down to escape sequences.

**Comment by mintty** [comment]
We've had such screenshots already above in this issue. You don't even say how mintty looked before it got into the displayed position and what you did to bring it there. That doesn't really help, despite your effort, sorry.
Escape sequences can be captured for example with the `script` tool, or with logging features of the terminals themselves.

**Comment by xelra** [comment]
The maximized windows were maximized by dragging them to the top of the screen. The half-maximized windows were half-maximized by dragging them to the far left of the screen. Those were the starting points. Cmder/ConEmu and PuTTY remained in those positions after starting Neovim, mintty didn't. Which can be seen in the screenshots.

I managed to capture some sequences with the script command. I'm not really able to make sense of those though. I hope it helps:
```
Script started on Sun 12 Mar 2017 12:58:55 AM CET
^[]0;user@ns397583:~^G^[[0;32muser^[[m ^[[1;34m~^[[m ^[[0;32m$^[[m nvim^M
^[[?1049h^[[H^[[2J^[[?2004h^[[?1004h^[[8;62;106t^[[H^[[2J^[[1;1H^[[1;1H^[[?12l^[[?25h^[[?25l^[[?1002h^[[?1006h^[[H^[[2J^[[1;1H^[[2;1H^[(B^[[m^[[94m~                                                                                                         ^[[3;1H~                                                                                                         ^[[4;1H~                                                                                                         ^[[5;1H~                                                                                                         ^[[6;1H~                                                                                                         ^[[7;1H~                                                                                                         ^[[8;1H~                                                                                                         ^[[9;1H~                                                                                                         ^[[10;1H~                                                                                                         ^[[11;1H~                                                                                                         ^[[12;1H~                                                                                                         ^[[13;1H~                                                                                                         ^[[14;1H~                                                                                                         ^[[15;1H~                                                                                                         ^[[16;1H~                                                                                                         ^[[17;1H~                                                                                                         ^[[18;1H~                                                                                                         ^[[19;1H~                                                                                                         ^[[20;1H~                                                                                                         ^[[21;1H~                                                                                                         ^[[22;1H~                                                                                                         ^[[23;1H~                                                                                                         ^[[24;1H~                                                                                                         ^[[25;1H~                                                                                                         ^[[26;1H~                                                                                                         ^[[27;1H~                                                                                                         ^[[28;1H~                                                                                                         ^[[29;1H~                                                                                                         ^[[30;1H~                                                                                                         ^[[31;1H~                                                                                                         ^[[32;1H~                                                                                                         ^[[33;1H~                                                                                                         ^[[34;1H~                                                                                                         ^[[35;1H~                                                                                                         ^[[36;1H~
```
I started the script, executed Neovim, closed Neovim and then got those 2 lines with `cat -v typescript`.

**Comment by mintty** [comment]
OK, the effective core of the sequence is the same as I quoted above, just with different values:
`echo -e '\e[8;62;106t'` , and we can reduce the line count e.g. to 30, for easier testing.
So if I just drag the window manually towards the left screen edge (no maximizing) and output that, the gap appears.

**Comment by xelra** [comment]
Yes, when I enter `echo -e '\e[8;62;106t'` it has exactly the same behavior.

> So if I just drag the window manually towards the left screen edge (no maximizing) and output that, the gap appears.
I move the window to the far left edge, which is a snap-point. That's when the window automatically resizes to exactly half the screen-size. You can do that on the right side and the left side. On the top edge the window will fully maximize.

I just wanted to explain that again, because I wasn't sure whether this was still unclear.

Steps to reproduce for the half-width-maximized window:
```
Move window to the far left and make it resize to the left half of the screen.
echo -e '\e[8;62;106t'   or start Neovim
```
I think the `\e[8;62;106t` depends on the current size of the window. Because for the full-width-maximized window, that sequence does not replicate the behavior.

Thanks for still keeping up with this issue.

**Comment by mintty** [comment]
This is apparently due to a change introduced in 2.1.1:
  * Trying to keep window position within monitor work area on resizing (#79).
The interesting thing is (when tracing the current version) that, although the monitor coordinates start at 0 on the left border, the x coordinate of a window dragged to the left edeg is negative (-8), so mintty thinks the window is outside the screen and adjusts that. I'll see how to tweak that but the code itself is not useless, so I'd prefer to understand first why the coordinate is negative...

**Comment by mintty** [comment]
It only happens in recent Windows versions, but mintty now has a workaround.

**Comment by mintty** [comment]
Released 2.7.6.

**Comment by lf-** [comment]
Built myself a copy of 2.7.6. Fix confirmed. The window still changes but it's a damn sight better than what it was. Thanks @mintty!

**Comment by xelra** [comment]
I've tried the new release in Cygwin. Unfortunately the problem is only partially fixed.

When half-width-maximized on the left side, the window now properly retains the top left corner. The bottom right corner does not stay maxmized though and the window resizes.

When half-width-maximized on the right side, the window does neither retain the top left nor the bottom right corner. The window moves and resizes.

When full-width-maximized, the window retains the top left corner, but the size of the window decorator on the top changes. Also the bottom right corner moves and the window resizes to be bigger than the screen.

**Comment by mintty** [comment]
As I've disabled aero-snap and Windows seems to be hiding this setting meanwhile, I cannot reproduce the issue. For all cases, please describe them more clearly; I don't see any problems with a fully maximized window. Option `-s maxwidth` does not seem to work at all anymore, if you mean that.

**Comment by xelra** [comment]
Here is a proper illustration of the current state.

**Half-width-maximized on the left side**
Before starting Neovim or running `echo -e '\e[8;62;106t'`
![left-before](https://cloud.githubusercontent.com/assets/908568/25316659/b4a4a46c-286b-11e7-9baf-9bdb6d54b5c9.png)
After starting Neovim or running `echo -e '\e[8;62;106t'`
![left-after](https://cloud.githubusercontent.com/assets/908568/25316665/cc101ce4-286b-11e7-81eb-14b736d4a0ec.png)
Here the top left corner stays in position now, but the bottom right corner still doesn't.

**Half-width-maximized on the right side**
Before starting Neovim or running `echo -e '\e[8;62;106t'`
![right-before](https://cloud.githubusercontent.com/assets/908568/25316672/f70a7c14-286b-11e7-8c0f-e404e63a6ed2.png)
After starting Neovim or running `echo -e '\e[8;62;106t'`
![right-after](https://cloud.githubusercontent.com/assets/908568/25316681/13e40d1e-286c-11e7-8176-4ee5303da91d.png)
Here the top left corner moves to the left side a bit. Probably by the same offset that we had in the original issue on the left side too. The bottom right corner seems to move in a similar way, so that the resized window seems to be the same size as on the left side.

**Full-width-maximized**
Before starting Neovim or running `echo -e '\e[8;63;213t'`
![max-before](https://cloud.githubusercontent.com/assets/908568/25316700/7935a2cc-286c-11e7-82c6-eff0c54ab07d.png)
After starting Neovim or running `echo -e '\e[8;63;213t'`
![max-after](https://cloud.githubusercontent.com/assets/908568/25316701/7eb5d9e2-286c-11e7-9c03-d923d3996981.png)
Note that the code is different. I captured it with script when starting Neovim. The full-width maximized window has a slimmer window decorator. When starting Neovim, the window seems to leave this mode and the size of the window decorator changes. Also the bottom right corner of the window moves. The top left corner seems to stay in place, similar to the working fix for the half-width-maximized window on the left side.

Thank you very much for your continued work on this issue.

**Comment by mintty** [comment]
> the bottom right corner of the window moves

moves down, correct?

For all three cases, I assume the actual terminal size (`stty size`) is not affected?

**Comment by xelra** [comment]
Yes, I've checked `stty size`. It's unchanged before and after. It's `62 106` for half-width-maximized and `63 213` for full-widht-maximized on my computer.

> moves down, correct?

As you can see from the screenshots, it shrinks the window for the half-width-maximized scenarios and grows the window in the full-width-maximized scenario.

**Comment by mintty** [comment]
I see two issues:
* the title bar gets larger (this makes of course the whole window larger accordingly, moving down the bottom border)
* the right-maximized window skips a bit left

The "half-maximized" windows have some additional padding at the bottom because the pixel size does not divide evenly by the character height. When resizing, this is adjusted to uniform padding, that's normal.

**Comment by xelra** [comment]
> The "half-maximized" windows have some additional padding at the bottom because the pixel size does not divide evenly by the character height. When resizing, this is adjusted to uniform padding, that's normal.

Other terminals don't share this behavior. The window doesn't resize, the remainder of the non-natural columns and rows stays as padding, instead of margin. Meaning it's inside the window border, not outside.

The problem with the resizing window is that it breaks features of the window manager. For example one can drag and drop the vertical line between 2 half-width-maximized windows and adjust the horizontal split from initially 50%/50% to e.g. 30%/70%. Then when using the snap features on additional windows, those will also maximize to either 30% or 70%, depending on which side you snap them to. It's a whole workflow that's ideal for widescreen monitors.

**Comment by mintty** [comment]
Setting max size is something different than setting a specific terminal size (in characters), for which the window size adjusts. Even if surprising in this case, it is not inconsistent behaviour.

Which window manager feature is broken? You can still drag the split border, right?

**Comment by xelra** [comment]
Yes, it's draggable, but not after the window moved or resized.

**Comment by mintty** [comment]
Please instruct me how to perform this connected dragging, so I have a reproducible test case.

**Comment by xelra** [comment]
I've made this gif to better illustrate the horizontal split functionality.
![loop](https://cloud.githubusercontent.com/assets/908568/25359366/365439bc-2945-11e7-8d5a-0f0899109950.gif)

**Comment by mintty** [comment]
If the terminal is resized to a size that it has already, window resizing is now skipped. Thus the bottom padding is kept, and the Windows magic around the aero-snapped window position should be retained.

The tweak I had introduced in 2.7.6 to be able to position on the left border does not seem to work anymore, so I've disabled it in the code. Windows has apparently changed its interpretation of window positions in incompatible ways; especially 0 is not 0 anymore. I'm not willing to further twist my mind about whether a workaround is possible for this crazy insanity. So also the "jumping" away from the right border will stay unchanged as long as Windows applies this silly new policy.

**Comment by mintty** [comment]
Released 2.7.7.

**Comment by xelra** [comment]
I've tried the new version. While the window retains its size now in the half-width scenarios, unfortunately it still jumps away from the border by a few pixels.

In the fully maximized scenario, neither size nor position stay unchanged.

Since this seems to be a really hard problem, I'd like to propose an option in the settings that lets you completely deactivate any automatic window re-positioning and -sizing.

Thank you for all the time you've already invested into this.

**Comment by mintty** [comment]
Not sure what operations you propose to prevent. I don't think Aero-snap rearrangement can be prevented (other than deactivating Aero-snap in the Windows system settings). All other rearrangements are not automatic. If it's possible to deactive Aero-snap per window class that might make sense. Please tell me how.

**Comment by xelra** [comment]
Sorry for the misunderstanding. I wasn't talking about aero-snap, but rather about window re-positioning and -sizing from inside the terminal. By means of escape sequences. To deactivate that with an option.

**Comment by mintty** [comment]
Resizing without changing the size was actually caught already, but not completely.
Fixed, and I assume this is sufficient. Suppressing the resize operation at all is not really compatible.

**Comment by mintty** [comment]
Released 2.7.8.

**Comment by mintty** [comment]
See also https://stackoverflow.com/questions/35259756/wpf-and-windows-10-invisible-border-around-windows.
